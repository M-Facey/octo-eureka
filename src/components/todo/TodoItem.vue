<script lang="ts" setup>
import TodoButton from "@/components/input/TodoButton.vue";

import IconCheck from "@/components/icons/IconCheck.vue";
import IconEdit from "@/components/icons/IconEdit.vue";
import IconClose from "@/components/icons/IconClose.vue";

import useScreenSize from "@/composables/useScreenSize";

export interface Prop {
  todoId: string;
  todoName: string;
  isCompleted: boolean;
  canEdit?: boolean;
  dataCy: string;
  hasSubTask?: boolean;
}

withDefaults(defineProps<Prop>(), {
  canEdit: false,
  hasSubTask: false,
});

defineEmits<{
  (e: "toggleCompleted"): void;
  (e: "editTodo", id: string): void;
  (e: "deleteTodo", id: string): void;
}>();

const { onMobile } = useScreenSize();
</script>

<template>
  <div
    class="w-full group flex items-center border bg-neutral-300 dark:bg-transparent dark:border-neutral-800 hover:border-transparent hover:bg-neutral-400 dark:hover:bg-black p-3 rounded-lg"
    :data-cy="dataCy"
  >
    <label
      for="taskCheck"
      class="relative w-6 h-6 flex items-center justify-center text-neutral-900 dark:text-white border border-neutral-500 rounded"
      :class="{ 'pointer-events-none': hasSubTask }"
      data-cy="todoCheck"
      @click="$emit('toggleCompleted')"
    >
      <input
        :id="todoId"
        type="checkbox"
        class="absolute w-full h-full opacity-0 peer"
        :checked="isCompleted"
      />
      <icon-check class="w-4 opacity-0 peer-checked:opacity-100" />
    </label>

    <p
      class="font-bold tracking-wide ml-2 mr-auto"
      :class="{
        'completed text-neutral-500': isCompleted,
        'text-neutral-900 dark:text-neutral-300': !isCompleted,
      }"
    >
      {{ todoName }}
    </p>

    <div
      class="flex gap-x-3 ml-auto transition-opacity"
      :class="{ 'opacity-0 group-hover:opacity-100': !onMobile }"
    >
      <todo-button
        v-if="canEdit"
        button-id="edit-todo"
        button-label="Edit Todo"
        button-size="sm"
        data-cy="editTodo"
        @trigger-event="$emit('editTodo', todoId)"
      >
        <icon-edit class="w-5" />
      </todo-button>
      <todo-button
        button-id="delete-todo"
        button-label="Delete Todo"
        button-size="sm"
        data-cy="deleteTodo"
        @trigger-event="$emit('deleteTodo', todoId)"
      >
        <icon-close class="w-5" />
      </todo-button>
    </div>
  </div>
</template>

<style scoped>
.completed {
  position: relative;
}

.completed::before {
  content: "";
  position: absolute;

  top: 50%;
  left: 0px;
  width: 100%;
  height: 2px;

  transform: translateY(-50%);
  @apply bg-neutral-900 dark:bg-neutral-300;
}
</style>
