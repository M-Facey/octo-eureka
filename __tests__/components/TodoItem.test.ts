import { mount } from "@vue/test-utils";
import { describe, it, expect } from "vitest";
import TodoItem from "@/components/todo/TodoItem.vue";

describe("Todo Item Component", () => {
  it("should render", () => {
    const wrapper = mount(TodoItem, {
      props: {
        todoId: "1",
        todoName: "Simple Task",
        isCompleted: true,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
