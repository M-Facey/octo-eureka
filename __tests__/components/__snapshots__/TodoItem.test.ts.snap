// Vitest Snapshot v1

exports[`Todo Item Component > should render 1`] = `
"<div data-v-549d9e09=\\"\\" class=\\"card group flex items-center border border-neutral-800 hover:border-transparent hover:bg-black mx-3 p-3 rounded-lg\\"><label data-v-549d9e09=\\"\\" for=\\"taskCheck\\" class=\\"relative w-6 h-6 flex items-center justify-center text-white border border-neutral-500 rounded\\"><input data-v-549d9e09=\\"\\" id=\\"1\\" type=\\"checkbox\\" class=\\"absolute w-full h-full opacity-0 peer\\"><svg data-v-549d9e09=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" class=\\"feather feather-check w-4 opacity-0 peer-checked:opacity-100\\">
      <polyline points=\\"20 6 9 17 4 12\\"></polyline>
    </svg></label>
  <p data-v-549d9e09=\\"\\" class=\\"font-bold tracking-wide ml-2 mr-auto completed text-neutral-500\\">Simple Task</p>
  <div data-v-549d9e09=\\"\\" class=\\"flex gap-x-3 ml-auto opacity-0 group-hover:opacity-100 transition-opacity\\"><button data-v-549d9e09=\\"\\" aria-label=\\"Delete Todo\\" class=\\"text-neutral-300 rounded flex gap-1 p-2 bg-neutral-800 hover:bg-neutral-700\\"><svg data-v-549d9e09=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" class=\\"feather feather-x w-5\\">
        <line x1=\\"18\\" y1=\\"6\\" x2=\\"6\\" y2=\\"18\\"></line>
        <line x1=\\"6\\" y1=\\"6\\" x2=\\"18\\" y2=\\"18\\"></line>
      </svg>
      <!--v-if-->
    </button></div>
</div>"
`;
